### PR TITLE
[WIP][HiveOnDelta]add hive-delta read

### PR DESCRIPTION
**Pull request by windpiger**
Implement HiveOnDelta Reader.
1. the HiveOnDelta subproject is put in external/hive-delta, other plugin(presto) can be put in external directory
2. Currently HiveOnDelta subproject is a independent  mvn project , maybe we should use sbt to manage it(I tried to use sbt subproject, but have not finished).
3. DeltaInputFormat is Implement , and there are some functions added in delta-core
4. DeltaInputFormat can support partition table /non-partition table , and the timetravel.
5. TestCase is not included in this WIP.
6. the schema in delta'metadata may be changed, so we should provide a tool to auto-create new external table with new schema 

```
create external table deltatblpar(region string,cnt string)
partitioned by(datee string)
row format serde 'org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe'
stored as inputformat 'io.delta.hive.DeltaInputFormat'
outputformat 'org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat'
location 'hdfs://emr-header-1:9000/delta/events';

// add partitions
msck repair table deltatblpar;
```
屏幕快照 2019-09-05 上午11 19 27

屏幕快照 2019-09-05 上午11 12 46

### CODE REVIEW

**Comment by databricks-cla-assistant** [comment]
[CLA assistant check](https://cla.dev.databricks.com/delta-io/delta?pullRequest=111) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.databricks.com/delta-io/delta?pullRequest=111) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dev.databricks.com/check/delta-io/delta?pullRequest=111) it.</sub>

**Comment by zsxwing** [comment]
Thanks a lot for your PR! This is awesome. I have several high level questions about this connector:

- Is it possible to make Hive use partition values in `AddFile`? https://github.com/delta-io/delta/blob/v0.3.0/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala#L111
- Since you are pulling Spark and all its dependencies, do you know which Hive versions are required? I assume some Hive versions may have conflicted dependencies.
- For the time travel support, is it possible to specify different versions for different tables in the same query?

**Comment by windpiger** [comment]
> Thanks a lot for your PR! This is awesome. I have several high level questions about this connector:
> 
> * Is it possible to make Hive use partition values in `AddFile`? https://github.com/delta-io/delta/blob/v0.3.0/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala#L111
> * Since you are pulling Spark and all its dependencies, do you know which Hive versions are required? I assume some Hive versions may have conflicted dependencies.
> * For the time travel support, is it possible to specify different versions for different tables in the same query?
Thank you for you reply~
* `AddFile` is the last code in the DeltaInputFormat(https://github.com/delta-io/delta/pull/111/files#diff-2345501dc817497d927a96a72992652fR67), and return FileStatus directly. I did not find where to use it in Hive
* Yes, there are lots of conflicted dependencies when build a shaded jar to use in Hive, especially janino and antlr4 related, Spark and Hive dependencies version is different. I test this use Hive 2.3.3/Hadoop2.7.2/Spark2.4.3. The resolved conflicted dependencies in this pom.xml is not suitable for other Hive version. I'd like to test more Hive version and provide corresponding pom.xml to resolve their conflicted dependencies.  One more that we should merge java META0INF.services  for DataSourceRegister by https://github.com/delta-io/delta/pull/111/files#diff-d65433519a94b76b3f98ef4acc583d2aR107 . Because json datasource is used in Delta, if not we will get the exception that there is no json DataSource
* Currently it does not support this in this WIP. Let me think about it how to support it. Maybe we can `set hive.delta.$tableName.versionAsOf=$version`, and use it in DeltaInputFormat.
And another scenario is that how do we guarantee that one table scanned by multiple times in one SQL query use the same snapshot. I think this logical should be put in the Hive engine not in the DeltaInputFormat.

**Comment by zsxwing** [comment]
@windpiger IIUC, the Delta partitioning doesn't work, and Hive will ignore it. Right?

**Comment by windpiger** [comment]
> @windpiger IIUC, the Delta partitioning doesn't work, and Hive will ignore it. Right?
Delta path is as the location of Hive Table or Hive table's partitions.
We can create a external Hive table from delta's metadata, or Alter table add partition location to Delta partition path.
Then we can use Hive to scan the Hive table/partition which backend is delta path.
If we select data from one Hive partition, the  dirs in DeltaInputFormat(https://github.com/delta-io/delta/pull/111/files#diff-2345501dc817497d927a96a72992652fR23) will include the delta partition path like '/delta/eventlog/a=1/b=2', then we can filter needed delta files in the partition path by Spark.

 So what we need is to provide a tool to auto-create hive table/partition from delta metadata.
and we can use it like other hive table.

**Comment by zsxwing** [comment]
Summary our offline discussion here:

We will look at the following non file format APIs to see if we can make Hive support Delta's partition:
https://cwiki.apache.org/confluence/display/Hive/StorageHandlers
https://cwiki.apache.org/confluence/display/Hive/FilterPushdownDev

**Comment by rtjarvis** [comment]
@windpiger - this is an awesome contribution. I'm playing with it now to add support for presto. I'm not able to get accurate results from Spark or Presto when updates have been made to the table. Is it functioning correctly?

**Comment by zsxwing** [comment]
@rtjarvis this PR doesn't add partitions automatically. That's why we are looking at [a low level API](https://cwiki.apache.org/confluence/display/Hive/StorageHandlers) to handle partitioning by ourselves.

**Comment by SpaceRangerWes** [comment]
@zsxwing I've been taking a look at the StorageHandler and how it has been implemented across a variety of applications. It is probably the right way to go, but the implementations of it have not been very consistent so it has been a challenge designing. 🙁

**Comment by windpiger** [comment]
> @windpiger - this is an awesome contribution. I'm playing with it now to add support for presto. I'm not able to get accurate results from Spark or Presto when updates have been made to the table. Is it functioning correctly?

Did you have implement PrestoOnDelta? Could you share the code and I can test it in my enviroment.
Could you provide more detail about how to update and how to get results from Spark or Presto?
Thanks!

**Comment by spmp** [comment]
>     * Yes, there are lots of conflicted dependencies when build a shaded jar to use in Hive, especially janino and antlr4 related, Spark and Hive dependencies version is different. I test this use Hive 2.3.3/Hadoop2.7.2/Spark2.4.3. The resolved conflicted dependencies in this pom.xml is not suitable for other Hive version. I'd like to test more Hive version and provide corresponding pom.xml to resolve their conflicted dependencies.  One more that we should merge java META-INF.services  for DataSourceRegister by https://github.com/delta-io/delta/pull/111/files#diff-d65433519a94b76b3f98ef4acc583d2aR107 . Because json datasource is used in Delta, if not we will get the exception that there is no json DataSource

I am trying with the Flokkr images so stuck with Hive `2.3.2` - As suspected give the above, when adding the _shaded_ JAR as:
```
ADD JAR /path/to/jar/delta-hive_2.11-1.0-SNAPSHOT.jar
```
And data source as:
```
create external table StreamDumpcdcstringa(time string,action string,name string,message string)
row format serde 'org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe'
stored as inputformat 'io.delta.hive.DeltaInputFormat'
outputformat 'org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat'
location 'hdfs://namenode:9000/delta/StreamDumpcdcstringa.delta';
```
When I go to get my data
```
select * from streamdumpcdcstringa;
...
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/spark/sql/delta/DeltaTableUtils$
	at io.delta.hive.DeltaInputFormat.listStatus(DeltaInputFormat.scala:30)
	at org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:322)
	at org.apache.hadoop.hive.ql.exec.FetchOperator.getNextSplits(FetchOperator.java:372)
```
I built Delta with Spark 2.4.3 and Hive 2.3.2 - Any chance you could help figure out how to shade the jar correctly for tihs combo? May be easier to use Hive 2.3.3

**Comment by windpiger** [comment]
> java.lang.NoClassDefFoundError: org/apache/spark/sql/delta/DeltaTableUtils
delta-core.jar should be also added to Hive.
You can set this by HIVE_AUX_JARS_PATH in hive-env.sh, like :
export HIVE_AUX_JARS_PATH=/path/to/hive-delta.ja,/path/to/delta-core.jar

**Comment by zsxwing** [comment]
Closing this. There is a new PR #241 for the Hive support.

File: external/hive-delta/src/main/scala/io/delta/hive/DeltaInputFormat.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  @throws[IOException]
+  override def listStatus(job: JobConf): Array[FileStatus] = {
+    val dirs = getInputPaths(job)
+    if (dirs.isEmpty) {
+      throw new IOException("No input paths specified in job")
+    } else {
+      TokenCache.obtainTokensForNamenodes(job.getCredentials, dirs, job)
+
+      // find delta root path
+      val rootPath = DeltaTableUtils.findDeltaTableRoot(spark, dirs)
```
Comment (Michael-YuSha) [inline external/hive-delta/src/main/scala/io/delta/hive/DeltaInputFormat.scala]: the master code have remove this method

**Comment by Michael-YuSha** [comment]
while i have create the hive table and location in aws 3， when i query the table by "select * fom delta_event limie 100; "i have got the err:
Exception in thread "main" com.google.common.util.concurrent.ExecutionError: java.lang.ExceptionInInitializerError
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2261)
	at com.google.common.cache.LocalCache.get(LocalCache.java:4000)
	at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4789)
	at org.apache.spark.sql.delta.DeltaLog$.apply(DeltaLog.scala:714)
	at org.apache.spark.sql.delta.DeltaLog$.forTable(DeltaLog.scala:676)
	at io.delta.hive.DeltaInputFormat.listStatus(DeltaInputFormat.scala:50)
	at org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:288)
	at org.apache.hadoop.hive.ql.exec.FetchOperator.getNextSplits(FetchOperator.java:372)
	at org.apache.hadoop.hive.ql.exec.FetchOperator.getRecordReader(FetchOperator.java:304)
	at org.apache.hadoop.hive.ql.exec.FetchOperator.getNextRow(FetchOperator.java:459)
	at org.apache.hadoop.hive.ql.exec.FetchOperator.pushRow(FetchOperator.java:428)
	at org.apache.hadoop.hive.ql.exec.FetchTask.fetch(FetchTask.java:147)
	at org.apache.hadoop.hive.ql.Driver.getResults(Driver.java:2208)
	at org.apache.hadoop.hive.cli.CliDriver.processLocalCmd(CliDriver.java:253)
	at org.apache.hadoop.hive.cli.CliDriver.processCmd(CliDriver.java:184)
	at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:403)
	at org.apache.hadoop.hive.cli.CliDriver.executeDriver(CliDriver.java:821)
	at org.apache.hadoop.hive.cli.CliDriver.run(CliDriver.java:759)
	at org.apache.hadoop.hive.cli.CliDriver.main(CliDriver.java:686)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.hadoop.util.RunJar.run(RunJar.java:234)
	at org.apache.hadoop.util.RunJar.main(RunJar.java:148)
Caused by: java.lang.ExceptionInInitializerError
	at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport$$anonfun$setSchema$2.apply(ParquetWriteSupport.scala:444)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport$$anonfun$setSchema$2.apply(ParquetWriteSupport.scala:444)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport$.setSchema(ParquetWriteSupport.scala:444)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat.buildReaderWithPartitionValues(ParquetFileFormat.scala:317)
	at org.apache.spark.sql.execution.FileSourceScanExec.inputRDD$lzycompute(DataSourceScanExec.scala:309)
	at org.apache.spark.sql.execution.FileSourceScanExec.inputRDD(DataSourceScanExec.scala:305)
	at org.apache.spark.sql.execution.FileSourceScanExec.inputRDDs(DataSourceScanExec.scala:327)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:627)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.UnionExec$$anonfun$doExecute$1.apply(basicPhysicalOperators.scala:571)
	at org.apache.spark.sql.execution.UnionExec$$anonfun$doExecute$1.apply(basicPhysicalOperators.scala:571)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:296)
	at org.apache.spark.sql.execution.UnionExec.doExecute(basicPhysicalOperators.scala:571)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.DeserializeToObjectExec.doExecute(objects.scala:89)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.MapPartitionsExec.doExecute(objects.scala:185)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.InputAdapter.inputRDDs(WholeStageCodegenExec.scala:391)
	at org.apache.spark.sql.execution.SerializeFromObjectExec.inputRDDs(objects.scala:110)
	at org.apache.spark.sql.execution.ProjectExec.inputRDDs(basicPhysicalOperators.scala:41)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:627)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.prepareShuffleDependency(ShuffleExchangeExec.scala:92)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:128)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.doExecute(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.InputAdapter.inputRDDs(WholeStageCodegenExec.scala:391)
	at org.apache.spark.sql.execution.SortExec.inputRDDs(SortExec.scala:121)
	at org.apache.spark.sql.execution.ProjectExec.inputRDDs(basicPhysicalOperators.scala:41)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:627)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.DeserializeToObjectExec.doExecute(objects.scala:89)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.MapPartitionsExec.doExecute(objects.scala:185)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.InputAdapter.inputRDDs(WholeStageCodegenExec.scala:391)
	at org.apache.spark.sql.execution.SerializeFromObjectExec.inputRDDs(objects.scala:110)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:627)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:80)
	at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:80)
	at org.apache.spark.sql.delta.util.StateCache$CachedDS.<init>(StateCache.scala:48)
	at org.apache.spark.sql.delta.util.StateCache$class.cacheDS(StateCache.scala:104)
	at org.apache.spark.sql.delta.Snapshot.cacheDS(Snapshot.scala:52)
	at org.apache.spark.sql.delta.Snapshot.<init>(Snapshot.scala:116)
	at org.apache.spark.sql.delta.DeltaLog$$anonfun$5.apply(DeltaLog.scala:152)
	at org.apache.spark.sql.delta.DeltaLog$$anonfun$5.apply(DeltaLog.scala:139)
	at scala.Option.map(Option.scala:146)
	at org.apache.spark.sql.delta.DeltaLog.<init>(DeltaLog.scala:139)
	at org.apache.spark.sql.delta.DeltaLog$$anon$3$$anonfun$call$1$$anonfun$apply$9.apply(DeltaLog.scala:718)
	at org.apache.spark.sql.delta.DeltaLog$$anon$3$$anonfun$call$1$$anonfun$apply$9.apply(DeltaLog.scala:718)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$.allowInvokingTransformsInAnalyzer(AnalysisHelper.scala:194)
	at org.apache.spark.sql.delta.DeltaLog$$anon$3$$anonfun$call$1.apply(DeltaLog.scala:717)
	at org.apache.spark.sql.delta.DeltaLog$$anon$3$$anonfun$call$1.apply(DeltaLog.scala:717)
	at com.databricks.spark.util.DatabricksLogging$class.recordOperation(DatabricksLogging.scala:77)
	at org.apache.spark.sql.delta.DeltaLog$.recordOperation(DeltaLog.scala:645)
	at org.apache.spark.sql.delta.metering.DeltaLogging$class.recordDeltaOperation(DeltaLogging.scala:103)
	at org.apache.spark.sql.delta.DeltaLog$.recordDeltaOperation(DeltaLog.scala:645)
	at org.apache.spark.sql.delta.DeltaLog$$anon$3.call(DeltaLog.scala:716)
	at org.apache.spark.sql.delta.DeltaLog$$anon$3.call(DeltaLog.scala:714)
	at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4792)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3599)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2379)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2342)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2257)
	... 24 more
Caused by: org.apache.parquet.schema.InvalidSchemaException: A group type can not be empty. Parquet does not support empty group without leaves. Empty group: spark_schema
	at org.apache.parquet.schema.GroupType.<init>(GroupType.java:92)
	at org.apache.parquet.schema.GroupType.<init>(GroupType.java:48)
	at org.apache.parquet.schema.MessageType.<init>(MessageType.java:50)
	at org.apache.parquet.schema.Types$MessageTypeBuilder.named(Types.java:1256)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetSchemaConverter$.<init>(ParquetSchemaConverter.scala:566)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetSchemaConverter$.<clinit>(ParquetSchemaConverter.scala)

**Pull request closed without merging**